pkg/virt-controller/watch/application.go

```diff
@@ -173,31 +199,58 @@ func Execute() {
 
 	app.migrationInformer = app.informerFactory.VirtualMachineInstanceMigration()
 
-	cache.WaitForCacheSync(stopChan, configMapInformer.HasSynced)
-	app.clusterConfig = virtconfig.NewClusterConfig(configMapInformer, app.kubevirtNamespace)
-
-	if app.clusterConfig.DataVolumesEnabled() {
+	if app.hasCDI {
 		app.dataVolumeInformer = app.informerFactory.DataVolume()
-		log.Log.Infof("DataVolume integration enabled")
+		log.Log.Infof("CDI detected, DataVolume integration enabled")
 	} else {
 		// Add a dummy DataVolume informer in the event datavolume support
 		// is disabled. This lets the controller continue to work without
 		// requiring a separate branching code path.
 		app.dataVolumeInformer = app.informerFactory.DummyDataVolume()
-		log.Log.Infof("DataVolume integration disabled")
+		log.Log.Infof("CDI not detected, DataVolume integration disabled")
 	}
 
 	app.initCommon()
 	app.initReplicaSet()
 	app.initVirtualMachines()
 	app.initDisruptionBudgetController()
 	app.initEvacuationController()
-	app.Run()
+	go app.Run()
+
+	select {
+	case <-app.reInitChan:
+		cancel()
+	}
+}
+
+// Detects if a config has been applied that requires
+// re-initializing virt-controller.
+func (vca *VirtControllerApp) configModificationCallback() {
+
+	vca.reInitLock.Lock()
```

Comment:
If you enqueue an element below, instead of closing, you will need no lock. `HasDataVolumeAPI` does a list operation on  the store. The store is already thread save.

Reply:
fixed